docs/adr/0002-shopping-lists-lineitem-and-textlineitem-update-actions.md

```diff
@@ -0,0 +1,336 @@
+          "textField": "text-1"
+        }
+      },
+     "addedAt": "2020-11-04T09:38:35.571Z"
+    },
+    {
+      "id": "24de3821-e27d-4ddb-bd0b-ecc99365285f",
+      "variant": {
+        "sku": "SKU-2"
+      },
+      "quantity": 2,
+      "custom": {
+        "type": {
+          "id": "4796e155-f5a4-403a-ae1a-04b10c9dfc54",
+          "obj": {
+            "key": "custom-type-for-shoppinglists"
+          }
+        },
+        "fields": {
+          "textField": "text-2"
+        }
+      },
+     "addedAt": "2020-11-04T09:40:12.341Z"
+    }
+  ]
+}
+</pre>
+        </td>
+    </tr>
+    <tr>
+        <th colspan="2">Analysis</th>
+    </tr>
+    <tr>
+        <td colspan="2">
+            <p>Draft has line items with <b>SKU-1</b> and <b>SKU-3</b> also in target project line item with
+                <b>SKU-1</b> exists, so <b>SKU-3</b> is a new line item, and <b>SKU-2</b> needs to be removed.</p>
+            <p> So we need to create an <a href="https://docs.commercetools.com/api/projects/shoppingLists#add-lineitem">AddLineItem</a> action
+                , a <a href="https://docs.commercetools.com/api/projects/shoppingLists#remove-lineitem">RemoveLineItem</a> action and
+                a <a href="https://docs.commercetools.com/api/projects/shoppingLists#change-lineitems-order">Change LineItems Order</a> 
+                of the line items <b>SKU-1</b> and <b>SKU-3</b>, because when we add line item with <b>SKU-1</b>
+                the order will be <b>SKU-1</b> and <b>SKU-3</b>.</p>
+            <p>The <b>challenge</b> in here is, those actions can not be added in one request because we don't know the
+                line item id of the new line item
+                with <b>SKU-3</b>, so we need to find another way to create a new line item with the right order.</p>
+            <p>Also another <b>challenge</b> in here is about the line item with <b>SKU-1</b>, as the order and data is
+                exactly same, we need to find a better way to avoid creating unnecessary actions.
+            </p>
+        </td>
+    </tr>
+    <tr>
+        <th colspan="2">Proposed solution</th>
+    </tr>
+    <tr>
+        <td colspan="2">
+            <p>
+                The solution idea about the new line item and changed order is still same with the case-1, but do we
+                need to remove and add line item with <b>SKU-1</b>? Which is not needed and we could start the removing
+                and adding from the changed order.
+            </p>
+            <pre lang="json">
+{
+  "version": 1,
+  "actions": [
+    {
+      "action": "removeLineItem",
+      "lineItemId": "24de3821-e27d-4ddb-bd0b-ecc99365285f"
+    },
+    {
+      "action": "addLineItem",
+      "sku": "SKU-3",
+      "quantity": 3,
+      "addedAt": "2020-11-05T10:00:10.101Z",
+      "custom": {
+        "type": {
+          "key": "custom-type-for-shoppinglists"
+        },
+        "fields": {
+          "textField": "text-3
+        }
+      }
+    }
```

Comment:
I suppose it is going to remove existing 'sku-2' line item.

These two updateActions are appropriate. I am just curious how we determine these two updateActions are needed in `UpdateActionUtil`. For example, if new line items is in order ['sku-3', 'sku-1'], we have to remove sku-1 from existing line items as well.

Reply:
Good catch!, You are right, the scenario was the same as scenario 1. https://github.com/commercetools/commercetools-sync-java/pull/614/commits/97d4d150d6872dec95395a741304e018303bc020

Reply:
From your example, you checked sku-3 and sku-4 is different than to able to match the order, all actions need to be removed and added back because you need to ensure the order.

#### In scenario 1, 
draft : [sku1,..., sku2]
target: [sku2...]

#### In scenario 2, 
draft: [sku1,..., sku3, sku2]
target: [sku1,..., sku2]

sku1 is exactly the same until you find the first difference assume you have N same but a difference in the middle, you don't need to create actions for those N items, here how we will know the difference depends on the comparisons between draft and target objects, so it's like doing sequence comparisons with their order. During the implementation, I could find more examples and could add here.